### DISCUSSION TITLE
[J4] Set minimum PHP to 7.4

### DISCUSSION THREAD

**Discussion by krileon** [joomla/joomla-cms#31881] [2. Cold Features]
PHP 7.2 is already EOL. PHP 7.3 will be EOL this year. PHP 7.4 has nearly 2 years of support planned. I think Joomla 4 shouldn't be releasing with an EOL minimum PHP. PHP 7.4 brings a lot of great features that I think would be of good use to Joomla itself.

https://www.php.net/supported-versions.php
https://www.php.net/manual/en/migration74.new-features.php

The following features specifically would be of significant use.

1. https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.typed-properties
2. https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.arrow-functions
3. https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.null-coalescing-assignment-operator
4. https://www.php.net/manual/en/migration74.new-features.php#migration74.new-features.core.unpack-inside-array
5. https://www.php.net/manual/en/migration73.new-functions.php

What are everyone's thoughts on this?

**Comment by brianteeman** [comment]
https://developer.joomla.org/about/stats.html

**Reply by krileon** [reply]
That has no bearing on PHP version availability. PHP 7.4 is already deployed on hosts and ready for anyone to move to. If you want better statistics use Wordpress for more idea on what budget hosting is providing to better judge availability. It's not outlandish to expect people to update their environment when moving to a new major release.

https://wordpress.org/about/stats/#php_versions

Releasing with a minimum PHP that's already EOL is dangerous. 7.2 will not get security fixes anymore. At the very least PHP 7.3 should be used, but it will be EOL before end of this year. Additional statistics below, which more accurately show PHP market trending towards 7.4.

https://w3techs.com/technologies/details/pl-php
https://w3techs.com/technologies/details/pl-php/7

**Comment by ghost** [comment]
Yes set J4 minimum PHP to 7.4.

krileon please also set minimum PHP to 7.4 to your extensions

**Comment by laoneo** [comment]
I'm bold now and do throw in:
Set it to 8, then we are safe for the next three years.

**Reply by ghost** [reply]
Set minimum PHP to 7.4 is ok, no need to set it to 8 now, why would you?

**Reply by laoneo** [reply]
Why would you not? PHP 7.4 will be in security support only mode, short after Joomla 4 is released. When Joomla 3 reaches end of life, we have to support a end of life PHP version again and we are in the same position as with Joomla 3 and PHP 5.3, just on Joomla 4, but much more earlier. I would only go with PHP 7.4 if our BC promise doesn't include to change the min PHP version during the lifecycle of a Joomla major version.

**Reply by infograf768** [reply]
The problem is that php8 is not available yet for all hosts

**Reply by laoneo** [reply]
Pretty sure, they are when Joomla 4 stable will be released.

**Reply by wilsonge** [reply]
I'm pretty sure that either you move to PHP 8 because of the union types or you stick with the current version which is based on the user data. This was something that came up on Production last week and we were going to look at the latest few months of stats data to see what people were running too

**Reply by laoneo** [reply]
That's exactly the issue. Many libraries will abandon PHP 7 pretty soon because of the incompatibility in syntax (example: https://github.com/Ocramius/CodeGenerationUtils/commit/02aab6ab303c93f60915d0f6d4ffaf0381a6671b) and not features. It is not the same situation like "do we go with PHP 7.2 or 7.3". There we can always lift the min PHP version during the lifecycle of Joomla 4, but not up to PHP 8. So we will have to manually patch our dependencies again in a year or two as we are forced to stay on PHP 7.x.

**Reply by ghost** [reply]
why not PHP 7.4 for its version lifetime only so PHP 7.4 support until its EOL ? So we will not have to patch our dependencies?

**Reply by laoneo** [reply]
Sorry, but I do not get your question.

**Reply by ghost** [reply]
Sorry, but what you do not get in my question?

**Comment by beat** [comment]
By the time Joomla 4 is released, PHP 8.0 will be available on all hosts... :wink: 

PHP 8.0 brings many modern language additions, most importantly [named arguments, union types, attributes, constructor property promotion and consistency](https://www.php.net/releases/8.0/en.php). Why not be bold ?

PHP 7.4 is really the bare minimum for credibility among developers. Joomla's success is linked to excitement for developers and users. We shouldn't be living in the past and looking to the future in the rear-view mirror. :wink: 

Reasonably, we could release Joomla 4.0 with PHP 8.0 recommended, and a minimum of PHP 7.4 for its version lifetime only (so PHP 7.4 support until its EOL).

When PHP 7.2 minimum was decided years ago, it made sense. Not anymore. And there is no development work involved to increase the minimum version to only non-EOL PHP versions.

With Joomla major versions tending to stay for very very long times, it would make sense to have dependencies (PHP/MySQL/Javascript/CSS/Webserver) support policy only for non-EOL versions. That way, e.g. Joomla 4.0.20 could require PHP 8.0, then 4.2.3 could require PHP 8.1, and so on. People not upgrading their PHP version would not be able to upgrade their Joomla, and would stay with a funcitoning website with a big red warning in the admin aera.

**Proposals:**

- Set minimum to non-EOL PHP version and recommended to latest stable. Probably/hopefully PHP 7.4 as minimum and 8.0.latest at Joomla 4.0 release date
- Publish a policy of ongoing PHP versions support corresponding to the versions' lifetimes
- Optionally, include a PHP version notice or warning depending on PHP versions run
- Add PHP minimum version information/check to package/pre-installer/version checker

**Reply by b2z** [reply]
Awesome proposal :+1:

**Reply by ali-95** [reply]
Awesome 👍

**Reply by nickanti** [reply]
totally agree

**Comment by alikon** [comment]
we can and we should do  like we have already stated for mysql

> 	MySQL 5.6 (Note: The minimum MySQL version can be subject to change during the lifetime of Joomla 4 if necessary)

PHP minimum requirements
PHP 7.4 (Note: The minimum PHP version can be subject to change during the lifetime of Joomla 4 if necessary)

**Reply by ghost** [reply]
why not PHP 7.4 for its version lifetime only so PHP 7.4 support until its EOL ?

**Comment by laoneo** [comment]
Symfony 6 goes with PHP 8.0
https://github.com/symfony/symfony/issues/40389

**Reply by brianteeman** [reply]
we might as well set it to 9.0 - maybe j4 will be released while that is still supported

**Comment by krileon** [comment]
Have we come to some form of consensus on what to do here? PHP 7.4 is a safe bet, but PHP 8.0 would be as well at this point. I think either would be great. The benefits of PHP 8.0 features on the future of the code base could be huge though. I've checked several major budget hosts and they're providing PHP 8.0 already so shouldn't be a concern regarding hosting at least.

**Comment by b2z** [comment]
PHP 7.4 should be the minimal version, that's for sure! There should be no discussion about it.

What we should discuss - are we brave enough to jump into PHP 8? 

Joomla 4 is ready for it, but extension developers  may be are far away from it. Or may be not, because there is no such big difference between PHP 7.2 and PHP 8.x. So any extension developer that wants to support Joomla 4 is already making its efforts to support PHP 7.2, so I think there will be no pain to support PHP 8 either.

**Comment by kernusr** [comment]
As an extension developer, I raise 2 arms and 2 legs in support of php 7.4
One of the most important things for me is typed for class properties

**Reply by progreccor** [reply]
totally agree

**Comment by dgrammatiko** [comment]
Personally, I think that the conversation (PHP min ver, MySQL min ver, javascript min ECMAScript ver, CSS min, etc) **should not** be just a static question "what should be the minimums for Joomla?".
There is a greater picture that most people seem to miss: projects/languages/even the WEB itself has switched to rolling updates. Fixing the problem for a specific point in time (eg the time that J4 will be released) is far less than half of the story that needs to be addressed here.  Basically what I'm trying to say here is that the production team needs a plan, I wrote another question that explains better the problem (and proposes a possible solution): https://github.com/joomla/joomla-cms/discussions/32784

**Reply by brianteeman** [reply]
php, apache and mysql have not moved to rolling updates

**Reply by dgrammatiko** [reply]
True, but checking their release cycles (esp for PHP) someone could see some similarities to NodeJS (js the language is on rolling updates since 2015)

PHP:
![Screenshot 2021-03-31 at 21 50 56](https://user-images.githubusercontent.com/3889375/113202794-81193180-926b-11eb-8f1b-109dfa775f8c.png)

Node:
![Screenshot 2021-03-31 at 21 51 20](https://user-images.githubusercontent.com/3889375/113202816-88403f80-926b-11eb-9146-60e6d0a5d336.png)

So, although you are technically right, there is evidence that even these are switching slowly and even if that is not what is happening there is the hard fact that PHP individual versions (as also NodeJS) have a 3 years lifespan. This, I guess somehow needs to be reflected in the product. I don't expect that J4 after 6 years will still support 7.3 (or whatever eventually will be the minimum)...

**Reply by krileon** [reply]
I simply think the PHP version should increase as needed. It shouldn't be frozen in time like past Joomla releases. This should especially be the case if a PHP version is no longer receiving security patches.

As is with PHP 7.2 Joomla 4 would be releasing with a literal EOL release of PHP. No security patches.. nothing.. **that is dangerous** from a security standpoint and Joomla should promote strong security.

PHP 7.4 gives Joomla until 2022 to move to PHP 8. I would prefer PHP 8 now, but can understand the hesitation from whomever is in charge of such a change since usage statistics for PHP 8 basically are non-existent. I also think Joomla should utilize the PHP version it sets instead of just bumping up a number so when Joomla does move to PHP 8 minimum it should use PHP 8 features to improve the core code.

**Comment by korenevskiy** [comment]
@krileon @dgrammatiko  @brianteeman @progreccor @infograf768  @alikon 
But interestingly, Joomla 4 is created for a long time. Here the main quality. But since this has been happening for a long time, do developers need to support PHP v 7?
For example, for old compatibility, you can raise the requirement to 7.4.
I want to note that in this question there is no topic, what will happen to users who used Joomla 4 on the old version of PHP. These users do not exist, since Joomla 4 is completely redesigned and it does not need to be supported.There are no longer those who already use it in combat mode.
On the other hand, you can talk about extension support, but every extension manufacturer should also think about PHP8 support when supporting J4. If any extension does not support PHP 8, then it is safe to say that such an extension does not support J4.
It has already been more than six months since PHP 8 was released. This is a long time for the programming world, and even more so for web programming. In 5 months, the community is already waiting for PHP 8.1.

[https://developer.joomla.org/about/stats.html](https://developer.joomla.org/about/stats.html)
This link shows that most sites are running PHP 5.6. Of course, this is true. After all, we should not look at the statistics of existing sites. We should look at the statistics of the created sites for this period this year.
**What is the ratio of sites in PHP 5.6 created this year? 
Is it possible that half of all new sites are now made in PHP 5.6?**
Let's look at the statistics of created sites on Joomla 4 (beta/dev/rc) with PHP 5.6. After all, according to statistics 5.6 is the most popular version of PHP. 0% !.

Sites created on 5.6 remain on that PHP because extensions installed (purchased) for that period were not updated. Those who can switch the PHP version can not do this, since the code is not compatible.
On Joomla 4, it is in a new format. Which means that there are no Joomla 4 sites that require an older version of PHP, because Joomla 4 is not out yet.
.
Well, now let's assume that in 2 years Joomla4 will not be able to use PHP 9. For the reason that Deprecated functions in PHP 8 will disappear completely in PHP 9, and for user compatibility, Joomla 4 will be stuck in PHP 8 for always. And so you will have to reinvent the wheel by creating Joomla 5. We'll have to come up with a migration wheel from Joomla 4 to Joomla 5.

Each of you what version of PHP will you use for Joomla 4, after release? 
1. PHP 5.6
2. PHP 7.2
3. PHP 7.4
4. PHP 8

I suggest you consider switching to PHP 8.
The concept of Joomla 4 is to get out of the quagmire of old technologies and use modern technologies. What for? Because that's what Joomla 4 was created for. But then why drag along the old technologies we're trying to get out of?

**Reply by krileon** [reply]
People don't upgrade their PHP version unless they have to. That's why the Joomla usage statistics are useless. Availability statistics are more important, which I've already posted above.

Joomla 4 should be releasing with a PHP 7.4 minimum. That gives Joomla till the end of 2022 to bump its minimum to PHP 8, which Joomla should absolutely be doing. Basically Joomla should be moving its requirements throughout its release cycle instead of staying static. As I've already explained releasing a new major stable release with EOL minimums IMO is dangerous and a security issue that should be taken into consideration.

As an extension developer our minimum for Community Builder will be PHP 7.4. We've already made this announcement and put a banner in CB to notify everyone that next stable release will have a minimum requirement of PHP 7.4. We will move this to PHP 8 when PHP 7.4 is EOL as we are taking a stance to no longer support EOL environments.

Alas this discussion seams to have bared no fruit and Joomla is going to release with PHP 7.2 anyway.

**Comment by laoneo** [comment]
Guess this one can be closed as Joomla 4 is shipped with 7.2.5.
https://downloads.joomla.org/technical-requirements

**Reply by beat** [reply]
Not really to be closed. All argument for and against in here are still valid. Sooner or later, Joomla will need to require a newer PHP, like PHP 7.4.

Actually, we could decide on a policy for minimum PHP versions that could close such discussions.
Also there is no reason to set in stone PHP minimum version for Joomla 4.XX. What should the minimum version be for Joomla 4.1, due in 5 months ? Decisions well in advance allow to put advanced "upgrade your PHP to a maintained version" messages in Joomla.

**Reply by laoneo** [reply]
I didn't say delete, just close. If you want to have a discussion about changing the policy to define new technical requirements during a major, then I would start a new topic and leave this one here as it is.
I remember we were once discussing the issue to change min PHP version during a major release, but can't remember where.

**Comment by sozzled** [comment]
I agree PHP7.4 minimum PHP version for J! 4; it makes sense.  I completely disagree that PHP 8—used on ~1% of all J! websites today—should be the "entry level".  PHP 8 (as the minimum standard) makes no sense ... unless to disincentivise J! 4 consumer take-up.

**Reply by laoneo** [reply]
As I wrote just before, it makes no sense to discuss a new min PHP version for Joomla 4 as long as we have the policy of a fixed technical requirement during a major release.

**Reply by korenevskiy** [reply]
And for 4.1 is there an opportunity to switch to 7.4?

**Reply by krileon** [reply]
> As I wrote just before, it makes no sense to discuss a new min PHP version for Joomla 4 as long as we have the policy of a fixed technical requirement during a major release.

That policy needs to be changed. It's an ancient policy that doesn't fit in todays technology. Any host that has 7.2 undoubtedly has 7.4.

We've already announced Community Builder will be requiring 7.4 in its next minor release. I encourage many other extension developers to do the same. Typed properties in classes alone are worth the switch, but I've also been making significant use of Null coalescing assignment operators.

> And for 4.1 is there an opportunity to switch to 7.4?

IMO 4.1 should make the change to 7.4.

**Reply by brianteeman** [reply]
>  Any host that has 7.2 undoubtedly has 7.4.

that is just guesswork

**Reply by krileon** [reply]
> > Any host that has 7.2 undoubtedly has 7.4.
> 
> that is just guesswork

No. No it's not. I would've agreed if I said that over a year ago, but now? Now it's fact. Usage statistics even reinforce this fact, but sure lets keep using a deprecated and unsupported PHP version with less features because that makes sense.

https://wordpress.org/about/stats/#php_versions
https://w3techs.com/technologies/details/pl-php/7

**Reply by sozzled** [reply]
On re-thinking this discussion, I don't have as much sympathy for changing the minimum version of PHP about PHP 7.3. In fact, I don't have much sympathy for changing the PHP entry level from PHP 7.3 for J! 4._anything_.  At this early stage of take-up by people who use J! for their websites, there is a cohort of users who have J! 3._x_ websites—sites that may or may not be updated or migrated to J! 4._x_ in the near future—and for many of those people it may prove impractical to force them to change to PHP 7.4 (or, worse, to PHP 8) while they remain dependent on PHP 7.3.  Even though this is not an issue for me personally, some people will be affected if they can't run PHP 7.3 within the same hosting environment as the one they're intending to use to update to J! 4.x (if the target [website] PHP version is incompatible with "source" [website]).  This is especially relevant to people, like me, who use a "shared hosting environment" to run several [J!] websites.

As I say, it's not something that I'm personally worried about.  I've spent the past six months or more evaluating my need to keep outdated—-pre-PHP 7.3—extensions, fix them or replace them.  Progressively, I've resolved fatal errors, deprecated warning, etc. over that time.  I'm uncomfortable with using PHP 8 at the moment because changing to that environment could result in some of my J! 3.10 websites becoming inoperative; as I say, not because of J! 3.10, _per se_, but because of the extensions I may be using.  Further, while I am able to use _phpMyAdmin_ within my "internet" hosting environment, I'm not (yet) able to use that product with _Wampserver_ on my PC-hosted development area if I change to PHP 8 there.  Therefore, until Wampserver ships a PHP 8-compatible version of  _phpMyAdmin_, I'm a little bit stuck.

**Reply by krileon** [reply]
PHP 8 right now is of course not happening, but there is little to no reason not to move on to 7.4 in 4.1 or 4.2 even. Vast majority of hosts were 7.4 ready over a year ago.

There's no real reason to go to 7.3 and not 7.4. 7.3 doesn't bring enough features that have significant use. Typed class properties in 7.4 is HUGE and everyone should be using it. This is also where Joomla table classes absolutely kill my soul if I'm being honest here.. they all use dynamic properties, which is awful.. they should all define their properties in their classes and do so with strict typing.

> Further, while I am able to use phpMyAdmin within my "internet" hosting environment, I'm not (yet) able to use that product with Wampserver on my PC-hosted development area if I change to PHP 8 there. Therefore, until Wampserver ships a PHP 8-compatible version of phpMyAdmin, I'm a little bit stuck.

Suggest moving away from Wampserver to at the very least Laragon where you can support multiple environments better. Ideally switch to Docker.

**Reply by brianteeman** [reply]
and ofc community builder has a strong incentive to force this change as they require it in order to do what they need to do. I would have a lot more sympathy if cb had been contributing the same to joomla.

**Reply by sozzled** [reply]
Thanks, @krileon.  You don't have to convince me about the advantages of using PHP 7.4.  :smile:  I'm just a simple guy, a hobbyist/enthusiast and (despite what people may think) my life doesn't revolve around J!.  What I was trying to say (and I make little apology for my clumsiness and lack of expertise) is that there are legions of J! users who wouldn't know a Docker from a Laragon—or a Laravel from a Symfony let alone a simple _x_AMP_x_ stack—and wouldn't give two seconds' time to checking these things out.  I wouldn't even know where to start (and I have no intention of learning how to start).  So, thanks for your suggestion but any discussion about moving away from Wampserver is too technical for my poor brain to contemplate.

**Reply by krileon** [reply]
> and ofc community builder has a strong incentive to force this change as they require it in order to do what they need to do. I would have a lot more sympathy if cb had been contributing the same to joomla.

You thinking extension developers don't contribute to Joomla is a big reason why I don't think I'll ever be able to get along with you. You think Wordpress is successful by itself? We continue to provide Community Builder for free and have done so since Joomla existed. I'll never understand why you consistently insist on being so unprofessional to other developers, which I might add is an amazing way to get people to contribute to Joomla. Frankly your attribute is a big reason I don't directly contribute to Joomla.

I don't even know why you think we're incentivized to push PHP 7.4. there's no commercial incentive here. It's just to be able to write better code and give extension developers a better foundation to build upon.

**Reply by brianteeman** [reply]
Very many do. It's just a few that are conspicuous in their absence. The point I was trying, and failing, to make is that if instead of criticism of the core code, limited by PHP, you had worked with others to improve the core then we wouldn't be having this conversation.

**Reply by krileon** [reply]
> Very many do. It's just a few that are conspicuous in their absence. The point I was trying, and failing, to make is that if instead of criticism of the core code, limited by PHP, you had worked with others to improve the core then we wouldn't be having this conversation.

My criticism for typed table class properties can't be implemented without PHP 7.4. Hence discussion here about PHP 7.4. Modern developers want to write modern code. To do that they need modern requirements. Additionally I'm managing CB and 20+ other projects. I don't have the time to directly contribute and if I did I couldn't because Joomla doesn't allow PHP 7.4 features.

[A PR is already being proposed to try an add this](https://github.com/joomla/joomla-cms/pull/35247), but it's basically just trying to add some ORM layer which Joomla should just use Doctrine for the DBA and ORM instead of reinventing the wheel here. I'm not about to write a bunch of code using time I don't have just for the PR to be shot down because Joomla doesn't want to make the move to PHP 7.4. That's why we're having this discussion.

Additionally you act like someone not directly contributing means they're not allowed to have an opinion. I guess every customer I've ever had should just stop having an opinion then. After all our products are open sourced and they could just contribute the change.

**Comment by dgrammatiko** [comment]
I will repeat what I wrote before: the technical requirements need a serious review.

To get a grasp of the degree of nonsense that goes into the current policy let me explain it in a very easy example. Here we go:
- Joomla 4 released on 17 August 2021
- On that day the latest Chrome (and by extension all the chromium browsers, give or take) was 92 nad due to policy Joomla support extended to version 91 (latest 2 versions) ref: https://en.wikipedia.org/wiki/Google_Chrome_version_history
- Joomla 4.04 is planned for the end of October (10 weeks after 4.0.0)
- On that day the latest Chrome will be 96
- By the current rules Chrome 96 and Chrome 95 are officially supported. The reality is totally different though...

So to make this clear in 10 weeks the minimum browser support will be shifted significantly although that might not be what the project really wants from contributors and their proposed code or sending the wrong kind of message to the end-users. In short, the policy doesn't make any sense for nobody, the project needs to move (at least for the front end) to the minimum acceptable percentage according to feature support

**Reply by korenevskiy** [reply]
Close Discussion!

**Comment by krileon** [comment]
Can we discuss increasing the PHP version for Joomla 4.1 now?

https://stitcher.io/blog/php-version-stats-january-2022
https://w3techs.com/technologies/details/pl-php/7

PHP 7.2 and 7.3 are dead. There's no reason to not push the minimum to 7.4.

**Reply by brianteeman** [reply]
18% usage is hardly dead and these stats are for probably the higher end of sites

**Reply by brianteeman** [reply]
PS 4.1 is now in Beta 3 so its of course too late for 4,1

**Reply by HLeithner** [reply]
We follow semver which means the next increase of the PHP Version will be in 5.0. The minimum php version in 5 will likely be 8.1

**Reply by brianteeman** [reply]
There is nothing stopping you from using a higher minimum for Community Builder

**Reply by krileon** [reply]
PHP 7.2 and 7.3 are both EOL. For features and security fixes. This policy really needs to change.

https://www.php.net/supported-versions.php

This isn't about Community Builder, which will require 7.4 and we will continue to bump its requirements from here on to ensure safe environments for our users. This is about contributors being able to contribute better, safer, and cleaner modern code. New talent wants to work with modern code. Joomla needs new talent to grow.

> As a rule, the developer should be concerned about the minimum version of PHP only when he makes extensions for general sale. Perhaps this case.

I've no idea what you mean by this. Community Builder is FREE and has been since 2004 before Joomla was even Joomla. We've been dedicated to Joomla for a very long time. Our interests are in Joomla's future.

The constant refusal to hear out developers that've been with this platform since its birth is just going to continue to drive developers away. At the very least consider that the current versioning policies need serious change.

Anyway, I give up.

**Reply by brianteeman** [reply]
If we broke semver and broke 20% of sites and changed the minimum php version as you demand will that mean you start to contribute to joomla?

**Reply by krileon** [reply]
> If we broke semver and broke 20% of sites and changed the minimum php version

Why would they just suddenly break? Add a banner notice to backend in 4.1 that the minimum PHP version is changing. Then change it in 4.2 for example. Don't allow updating if they don't meet the minimum.

> as you demand

I'm not demanding anything. I'm opening a conversation to make changes to the current release policies. PHP moves faster now. JS/CSS/Browsers move faster now. The current policies do not make sense for the current web. This isn't a situation where we need to be stuck with the same PHP version for 2-3+ years anymore.

> will that mean you start to contribute to joomla?

I've contributed over 10 years of my life to building a community of over half a million users entirely around Joomla. Our team has also contributed directly to Joomla for a very long time over those years where possible. I've had it with you thinking extension developers do not contribute to Joomla. Stop. This does not make you look good.

**Reply by b2z** [reply]
Will change to minimum PHP 7.4 will break public CMS API?

**Reply by dgrammatiko** [reply]
NO, raising the PHP version will just gonna require a newer version for installation and running the app, so, the increased version itself will not introduce any breaks in the API.

**Reply by b2z** [reply]
So Semver is not the case? :)

**Reply by dgrammatiko** [reply]
PHP 7.2 and 7.4 are compatible (7.4 had some new features but nothing that was breaking b/c). The question here is more about hosting providers and of course how the project wants to move forward, as the current scheme is completely gibberish…

**Comment by nickanti** [comment]
Is there a way to refresh the stats on PHP usage on Joomla installs? Its been a year since this discussion started and we seem to be still "using" 1 year old stats.

**Reply by brianteeman** [reply]
This is always up to date
https://developer.joomla.org/about/stats.html

or if you want the raw data https://developer.joomla.org/stats/cms_php_version

**Reply by krileon** [reply]
Most of which are statistics from out of date installs. I don't see how that's useful. They're not going to update PHP just to update PHP. A new version requiring a newer PHP would get them to update it, but it's not like that's going to matter to the massive amount of 3.6 - 3.9 installs as they're not updating to begin with.

**Reply by korenevskiy** [reply]
> Most of which are statistics from out of date installs. I don't see how that's useful. They're not going to update PHP just to update PHP. A new version requiring a newer PHP would get them to update it, but it's not like that's going to matter to the massive amount of 3.6 - 3.9 installs as they're not updating to begin with.

There are two right ideas. 1. The client should work stably 5-7 years after the creation of the site. 2.CMS should also move forward, this will allow it to be competitive with other CMS,Frameworks. Idea 1 should be provided with security patches. Joomla 4.1 is not a security update, it is an update of new features. Therefore, if the Joomla community is ready to guarantee critical security patches for version 4.0 for 5 years. Then there are no barriers in Joomla 4.1 to switch to 7.4. It seems to me that we are constantly replacing the minimum sufficient requirements with a desire to please everyone.

**Reply by korenevskiy** [reply]
Please watch. https://developer.joomla.org/about/stats.html Here the statistics do not say anything at all. These are statistics to get into the Guinness Book of Records, but no more. Let's consider the case when in a corporation of 10,000 people, each boss receives a salary 10 times more than his 10 subordinates. And the main boss receives a salary equivalent to all salaries combined. As can be seen from the number of employees, there are only 5 levels of positions. If the statistics department will calculate the average salary. then each employee, as it were, will receive 5 times more than he actually gets money on his hands. It's the same here. These circles tell us that Joomla is ready to become a candidate for the Guinness Book of Records, no more. But we need a table or a multidimensional graph (3d) where we can see the ratio of Joomla versions to PHP version and quantity. 
@brianteeman 
Here at the specified link https://developer.joomla.org/stats/cms_php_version we see raw data, but what does the digit after the colon mean in each value?
This will be proof for both sides. And it will also be seen that PHP 7.4 does not affect those who sit on the old Joomla or it will be seen that PHP 7.2 really needs to be supported with the last of its strength in the new Joomla.
Since PHP 7.4 and 8 support will attract startups. I hear many times that Laravel is good, and Joomla Framework is more difficult to work with.
